providers/Ototoy/mod.ts

```diff
@@ -0,0 +1,460 @@
+			case 'artist': {
+				return new URL(`https://ototoy.jp/_/default/a/${entity.id}`);
+			}
+			case 'package': {
+				return new URL(`https://ototoy.jp/_/default/p/${entity.id}`);
+			}
+			case 'label': {
+				return new URL(`https://ototoy.jp/labels/${entity.id}`);
+			}
+		}
+
+		throw new ProviderError(this.name, `Unable to determine type of entity ID '${entity.id}'`);
+	}
+
+	override getLinkTypesForEntity(_entity: EntityId): LinkType[] {
+		return ['paid download'];
+	}
+
+	scrapePackage(html: string, webUrl: URL): PackagePage {
+		const doc = new DOMParser().parseFromString(html, 'text/html');
+		if (!doc) {
+			throw new ResponseError(this.name, `Failed to parse HTML`, webUrl);
+		}
+
+		const thumbUrl = this.parseAlbumArtwork(doc);
+		if (!thumbUrl) {
+			throw new ResponseError(this.name, `Failed to extract album thumbnail`, webUrl);
+		}
+
+		const albumMeta = this.parseAlbumMeta(doc);
+		if (!albumMeta) {
+			throw new ResponseError(this.name, `Failed to extract album metadata`, webUrl);
+		}
+
+		const trackList = this.parseTracklist(doc);
+		if (!trackList) {
+			throw new ResponseError(this.name, `Failed to extract tracklist`, webUrl);
+		}
+
+		return {
+			thumbUrl,
+			albumMeta,
+			trackList,
+		};
+	}
+
+	// The format is as follows:
+	//
+	// <div class="album-artwork">
+	//    <button name="artwork-modal" type="button">
+	//       <figure>
+	//          <img src="https://imgs.ototoy.jp/..." class="photo">
+	//       </figure>
+	//    </button>
+	// </div>
+	//
+	// This is just the small thumbnail, the full size image comes from getArtwork()
+	parseAlbumArtwork(doc: HTMLDocument): string | undefined {
+		const imageElement = doc.querySelector('div.album-artwork img.photo');
+		if (!imageElement) return undefined;
+
+		return imageElement.getAttribute('src') || undefined;
+	}
+
+	// The format is as follows:
+	//
+	// <table id="tracklist" class="tracklist">
+	//    <tbody>
+	//       <tr class="disc-row">
+	//          <td colspan="5">DISC 1</td>
+	//       </tr>
+	//       <tr class="">
+	//          <td class="num center">
+	//             1
+	//          </td>
+	//          <td class="item">
+	//             <span id="title-123456">Free As A Bird (1995 Mix - Remastered)</span>
+	//          </td>
+	//          <td class="item center">04:25</td>
+	//       </tr>
+	//    </tbody>
+	// </table>
+	//
+	// NOTE: `disc-row` is optional
+	parseTracklist(doc: HTMLDocument): Track[] | undefined {
+		const trackListRows = doc.querySelectorAll('#tracklist tbody tr');
+
+		let currentDisc = null;
+		const tracks: Track[] = [];
+
+		for (const trackRow of trackListRows) {
+			if (trackRow.classList.contains('disc-row')) {
+				const match = trackRow.textContent.match(/\d+/);
+
+				if (match) {
+					currentDisc = parseInt(match[0], 10);
+				}
+
+				continue;
+			}
+
+			const trackNumberCell = trackRow.querySelector('.num');
+			if (!trackNumberCell) continue;
+
+			const trackNumber = trackNumberCell.textContent?.trim() ?? '';
```

Comment:
In a couple of places:
```suggestion
			const trackNumber = trackNumberCell.textContent.trim();
```
This doesn't seem to be necessary according to TypeScript?

Reply:
Yeah, didn't think so. I think my IDE was just bugging